http/events.go

```diff
@@ -69,14 +88,88 @@ func handleEventsSubscribeWebsocket(args eventSubscribeArgs) (websocket.StatusCo
 				logger.Warn("Could not serialize websocket event", "error", err)
 				return 0, "", err
 			}
-			err = args.conn.Write(ctx, messageType, messageBytes)
+			err = sub.conn.Write(ctx, messageType, messageBytes)
 			if err != nil {
 				return 0, "", err
 			}
 		}
 	}
 }
 
+// allowMessageCached checks that the message is allowed to received by the websocket.
+// It caches results for specific namespaces, data paths, and event types.
+func (sub *eventSubscriber) allowMessageCached(message *logical.EventReceived) bool {
+	if sub.isRootToken {
+		// fast-path root tokens
+		return true
+	}
+
+	messageNs := strings.Trim(message.Namespace, "/")
+	dataPath := ""
+	if message.Event.Metadata != nil {
+		dataPathField := message.Event.Metadata.GetFields()[logical.EventMetadataDataPath]
+		if dataPathField != nil {
+			dataPath = dataPathField.GetStringValue()
+		}
+	}
+	if dataPath == "" {
+		// Only allow root tokens to subscribe to events with no data path, for now.
+		return false
+	}
+	cacheKey := fmt.Sprintf("%v!%v!%v", messageNs, dataPath, message.EventType)
+	_, ok := sub.checkCache.Get(cacheKey)
+	if ok {
+		return true
+	}
+
+	// perform the actual check and cache it if true
+	ok = sub.allowMessage(messageNs, dataPath, message.EventType)
+	if ok {
+		err := sub.checkCache.Add(cacheKey, ok, webSocketRevalidationTime)
+		if err != nil {
+			sub.logger.Debug("Error adding to policy check cache for websocket", "error", err)
+			// still return the right value, but we can't guarantee it was cached
+		}
+	}
+	return ok
+}
+
+// allowMessage checks that the message is allowed to received by the websocket
+func (sub *eventSubscriber) allowMessage(eventNs, dataPath, eventType string) bool {
+	// does this even match the requested namespaces
+	matchedNs := false
+	for _, nsPattern := range sub.namespacePatterns {
+		if glob.Glob(nsPattern, eventNs) {
+			matchedNs = true
+			break
+		}
+	}
+	if !matchedNs {
+		return false
```

Comment:
Do we ever expect this to return false given that we've already provided `sub.namespacePatterns` as a filter above while subscribing?

`sub.events.SubscribeMultipleNamespaces(ctx, sub.namespacePatterns, sub.pattern)`

Reply:
It's more of a double-check, but probably not strictly necessary. It helped in writing tests as well.